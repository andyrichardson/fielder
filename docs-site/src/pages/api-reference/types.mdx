# Types

## FieldState

_The state of a field

## UseFieldProps

_Props for a form-input element (`input`, `select`, etc.)._

### Properties

#### `name`

_Field name._

Type: `string`

#### `value`

_Field value._

Type: `FormValue`

#### `onChange`

_Change event handler._

Type: `ChangeEventHandler`

#### `onBlur`

_Blur event handler (causes field to be `touched`)._

Type: `() => void`

## UseFieldMeta

_Metadata about a field's state._

### Properties

#### `touched`

_Whether the field has been touched (see `onBlur`)._

Type: `boolean`

#### `error`

_Validation error on a field._

Type: `Error | string | undefined`

#### `isValid`

_Whether the field is passing validation._

Type: `boolean`

#### `isValidating`

_Whether the field is currently undergoing async validation._

Type: `boolean`

