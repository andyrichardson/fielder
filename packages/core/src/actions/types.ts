export type StoreDispatch<A, T = any> = (action: A) => T;
export type StoreResolver<T, A> = (
  store: T,
  action: A,
  dispatch: StoreDispatch<A, T>
) => T;
