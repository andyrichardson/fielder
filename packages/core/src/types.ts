import {
  SetFieldValueArgs,
  ValidateFieldArgs,
  MountFieldArgs,
  BlurFieldArgs,
  UnmountFieldArgs,
  SetFieldStateArgs,
  SetFieldValidationArgs,
  BlurFieldAction,
  MountFieldAction,
  SetFieldStateAction,
  SetFieldValidationAction,
  SetFieldValueAction,
  UnmountFieldAction,
  ValidateFieldAction,
  ValidateSubmissionAction,
} from './actions';

export type FormValue = string | boolean | number | string[];

export type FormSchemaType = Record<string, FormValue>;

export type FieldsState<T extends FormSchemaType = FormSchemaType> = {
  [K in keyof T]: FieldState<T, K>;
};

export interface FormState<T extends FormSchemaType = FormSchemaType> {
  fields: FieldsState<T>;
  /** All mounted fields are valid. */
  isValid: boolean;
  /** Async validation currently active on a mounted fields. */
  isValidating: boolean;
  /** Set value for a field. */
  setFieldValue: <K extends keyof T>(a: SetFieldValueArgs<T, K>) => void;
  /** Trigger blur event for a mounted field. */
  blurField: (a: BlurFieldArgs<T>) => void;
  /** Force trigger validation on a mounted field. */
  validateField: (a: ValidateFieldArgs<T>) => void;

  /** Internal: Premount field during render */
  premountField: <K extends keyof T & string>(
    a: MountFieldArgs<T, K>
  ) => FieldState<T, K>;
  /** Internal: Manually mount field. */
  mountField: <K extends keyof T & string>(
    a: MountFieldArgs<T, K>
  ) => FieldState<T, K>;
  /** Internal: Manually unmount field. */
  unmountField: (a: UnmountFieldArgs<T>) => void;
  /** Internal: Manually set field state. */
  setFieldState: <K extends keyof T>(a: SetFieldStateArgs<T, K>) => void;
  /** Internal: Set new field validation function */
  setFieldValidation: <K extends keyof T>(
    a: SetFieldValidationArgs<T, K>
  ) => void;

  /** Trigger submission validation.
   *
   * Throws on synchronous validation errors.
   * Returns promise if form contains async validation.
   */
  validateSubmission: () => MaybePromise<{
    state: FieldsState;
    errors: Record<string, string>;
  }>;
}

export interface FieldState<
  T extends FormSchemaType | FormValue = FormValue,
  K extends keyof T = any,
  // Resolve value and schema form generics
  V extends FormValue = T extends FormSchemaType ? T[K] : T,
  S extends FormSchemaType = T extends FormSchemaType ? T : FormSchemaType
> {
  /** The field is currently mounted. */
  readonly _isActive: boolean;
  /** Validation function. */
  readonly _validate?: ValidationFn<V, S> | ObjectValidation<V, S>;

  // Props
  /** Field name */
  readonly name: K;
  /** Field value */
  readonly value: V;

  // Meta
  /** Field error */
  readonly error?: string;
  /** Field is currently valid. */
  readonly isValid: boolean;
  /** Field is currently being validated (async). */
  readonly isValidating: boolean;
  /** Field has been blurred since mount. */
  readonly hasBlurred: boolean;
  /** Field has been changed since mount. */
  readonly hasChanged: boolean;
}

/**
 * Events which trigger validation
 *
 * `mount`: Field has been mounted.
 *
 * `blur`: Field has had 'onBlur' event.
 *
 * `change`: Field has had 'onChange' event.
 *
 * `update`: The value of another field in the form has changed.
 *
 * `submit`: Submission has begun.
 */
export type ValidationTrigger =
  | 'mount'
  | 'blur'
  | 'change'
  | 'update'
  | 'submit';

/** Arguments passed to a validation function */
export type ValidationArgs<
  T extends FormValue = FormValue,
  S extends FormSchemaType = FormSchemaType
> = {
  trigger: ValidationTrigger;
  value: T;
  form: FieldsState<S>;
};

/** Handler for validation event */
export type ValidationFn<
  T extends FormValue = FormValue,
  S extends FormSchemaType = FormSchemaType
> = (args: ValidationArgs<T, S>) => void | Promise<void>;

/** A map of validation events corresponding to a function. */
export type ObjectValidation<
  T extends FormValue = FormValue,
  S extends FormSchemaType = FormSchemaType
> = {
  [k in ValidationTrigger]?: ValidationFn<T, S>;
};

type MaybePromise<T> = T | Promise<T>;

export type Dispatch<T> = (a: T) => void;

export type SetStateAction<T> = T | ((a: T) => T);

export type FormAction =
  | BlurFieldAction
  | MountFieldAction
  | SetFieldStateAction
  | SetFieldValidationAction
  | SetFieldValueAction
  | UnmountFieldAction
  | ValidateFieldAction
  | ValidateSubmissionAction;

export type FormOpts = { fromState?: Record<string, any> };
