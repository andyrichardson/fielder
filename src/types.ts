import { SetFieldValueArgs, BlurFieldArgs, UnmountFieldArgs } from './useForm';

export type FormError = Error | string;

export type FormValue = string | boolean | number | string[];

export type FieldsState<T = any> = Record<
  keyof T & string,
  FieldState<T[keyof T & string]>
>;

export interface FormState<T extends Record<string, any> = any> {
  fields: FieldsState<T>;
  isValid: boolean;
  isValidating: boolean;
  mountField: (k: FieldConfig<T>) => void;
  unmountField: (k: UnmountFieldArgs<T>) => void;
  setFieldValue: (a: SetFieldValueArgs<T>) => void;
  blurField: (a: BlurFieldArgs<T>) => void;
  validateField: (a: { name: keyof T & string }) => void;
  validateFields: () => void;
}

export interface FieldState<T = string | boolean | number | string[]> {
  readonly _isActive: boolean;
  readonly _validateOnChange: boolean;
  readonly _validateOnBlur: boolean;
  readonly _validateOnUpdate: boolean;
  readonly _validate: FieldConfig['validate'];

  // Props
  readonly name: string;
  readonly value?: T;

  // Meta
  readonly error?: FormError;
  readonly touched: boolean;
  readonly isValid: boolean;
  readonly isValidating: boolean;
}

export interface FieldConfig<
  S = any,
  K extends (keyof S & string) | string = S extends Record<string, any>
    ? keyof S & string
    : string,
  V = S extends Record<string, any> ? S[K] : S,
  F = S extends Record<string, any> ? S : unknown
> {
  /** Unique identifier for field. */
  readonly name: K;
  /** Validation function (throws errors). */
  readonly validate?: (value: V, form: F) => void;
  /** When the given field's value changes. */
  readonly validateOnChange?: boolean;
  /** When the given field loses focus. */
  readonly validateOnBlur?: boolean;
  /** When any change is made to the form state (global). */
  readonly validateOnUpdate?: boolean;
  /** Starting value. */
  readonly initialValue?: V;
  /** Starting error. */
  readonly initialError?: FormError;
  /** Starting touched state. */
  readonly initialTouched?: boolean;
  /** Starting valid state. */
  readonly initialValid?: boolean;
  /** Should destroy value when useField hook is unmounted. */
  readonly destroyOnUnmount?: boolean;
}
