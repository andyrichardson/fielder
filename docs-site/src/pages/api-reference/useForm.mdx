# useForm

_Creates a new form instance._

```tsx
import { useForm } from 'fielder'; 
```

## Example Usage

```tsx
const formState = useForm();

return (
  <FielderProvider value={formState}>
    {children}
  </FielderProvider>  
)
```


## Response

_A `FormState` object (state of the form)._

`FormState`

## Arguments

`useForm` doesn't currently take any arguments.

