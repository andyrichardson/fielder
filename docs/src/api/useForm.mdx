---
name: useForm
menu: Api
route: /api/useform
---

# useForm

_Creates a new form instance._

```tsx
import { useForm } from 'fielder';
```

## Example Usage

```tsx
const formState = useForm();

return <FielderProvider value={formState}>{children}</FielderProvider>;
```

## Return type

_The state of the form [(FormState)](#formstate) along with accessors and mutators._

Type: `FormState`

## Arguments

_useForm_ doesn't take any arguments.

## Types

### FormState

The state of the whole form along with accessors and mutators.

```ts
export interface FormState<T extends Record<string, any> = any> {
  fields: Record<string, FieldState>;
  isValid: boolean;
  isValidating: boolean;
  mountField: (k: FieldConfig<T>) => void;
  unmountField: (k: UnmountFieldArgs<T>) => void;
  setFieldValue: (a: SetFieldValueArgs<T>) => void;
  blurField: (a: BlurFieldArgs<T>) => void;
  validateField: (a: { name: keyof T & string }) => void;
  validateFields: () => void;
}
```

### FieldState

The state of an individual field (including meta information).

```tsx
export interface FieldState<T = string | boolean | number> {
  // Internals
  readonly _isActive: boolean;
  readonly _validateOnChange: boolean;
  readonly _validateOnBlur: boolean;
  readonly _validateOnUpdate: boolean;
  readonly _validate: FieldConfig['validate'];

  // Props
  readonly name: string;
  readonly value?: T;

  // Meta
  readonly error?: FormError;
  readonly touched: boolean;
  readonly isValid: boolean;
  readonly isValidating: boolean;
}
```
