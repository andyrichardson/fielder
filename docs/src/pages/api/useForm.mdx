# useForm

_Creates a new form instance._

```tsx
import { useForm } from 'fielder';
```

## Example Usage

```tsx
const formState = useForm();

return <FielderProvider value={formState}>{children}</FielderProvider>;
```

## Return type

The state of the form [(FormState)](#formstate) along with accessors and mutators.

Type: `FormState`

## Arguments

_useForm_ optionally takes a single object with the following properties:

### fromState

A key value store of field names and their values, for restoring form state on initial render.

Type: `object`

Example: `{ name: "some name", someField: "1234" }`

## Types

### FormState

The state of the whole form along with accessors and mutators.

```ts
export interface FormState<T extends Record<string, any> = any> {
  fields: Record<string, FieldState>;
  isValid: boolean;
  isValidating: boolean;
  setFieldValue: (a: SetFieldValueArgs) => void;
  blurField: (a: BlurFieldArgs) => void;
  validateField: (a: { name: string }) => void;

  // Internals
  premountField: (a: MountFieldArgs) => void;
  mountField: (a: MountFieldArgs) => void;
  unmountField: (a: UnmountFieldArgs) => void;
  setFieldState: (a: SetFieldStateArgs)
}
```

### FieldState

The state of an individual field (including meta information).

```tsx
export interface FieldState<T = string | boolean | number> {
  // Internals
  readonly _isActive: boolean;
  readonly _validate: FieldConfig['validate'];

  // Props
  readonly name: string;
  readonly value?: T;

  // Meta
  readonly error?: FormError;
  readonly isValid: boolean;
  readonly isValidating: boolean;
  readonly hasBlurred: boolean;
  readonly hasChanged: boolean;
}
```
