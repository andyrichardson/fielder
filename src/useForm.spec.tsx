import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import React from 'react';
import { useForm } from './useForm';
import { FormState } from './types';

let response: FormState;

const Fixture = () => {
  response = useForm();

  return null;
};

beforeEach(jest.clearAllMocks);

describe('initial call', () => {
  it('matches snapshot', () => {
    mount(<Fixture />);
    expect(response).toMatchSnapshot();
  });
});

describe('on mount field', () => {
  beforeEach(() => {
    mount(<Fixture />);
  });

  describe('on default args', () => {
    beforeEach(() => {
      act(() => response.mountField({ name: 'test' }));
    });

    describe('field state', () => {
      it('matches snapshot', () => {
        expect(response.fields).toMatchSnapshot();
      });
    });
  });

  describe('on provided args', () => {
    beforeEach(() => {
      act(() =>
        response.mountField({
          name: 'test',
          initialError: 'some error',
          initialTouched: true,
          initialValue: 'hello',
          destroyOnUnmount: false,
          initialValid: true,
          validate: jest.fn(),
          validateOnBlur: false,
          validateOnChange: false,
          validateOnUpdate: true
        })
      );
    });

    describe('field state', () => {
      it('matches snapshot', () => {
        expect(response.fields).toMatchSnapshot();
      });
    });
  });
});

describe('on unmount field', () => {
  const fieldName = 'TargetField';
  const value = 5678;

  beforeEach(() => {
    mount(<Fixture />);
  });

  describe('form state', () => {
    beforeEach(() => {
      act(() => {
        response.mountField({
          name: 'test',
          initialValue: 1234,
          initialValid: true
        });
        response.mountField({
          name: fieldName,
          initialValue: value,
          initialValid: false
        });
        response.unmountField({ name: fieldName });
      });
    });

    it('is valid', () => {
      expect(response).toHaveProperty('isValid', true);
    });
  });

  describe('fields state', () => {
    beforeEach(() => {
      act(() => {
        response.mountField({
          name: 'test',
          initialValue: 1234,
          initialValid: true
        });
        response.mountField({
          name: fieldName,
          initialValue: value,
          initialValid: false
        });
        response.unmountField({ name: fieldName });
      });
    });

    it('matches snapshot', () => {
      expect(response.fields).toMatchSnapshot();
    });

    it('is not active', () => {
      expect(response.fields[fieldName]).toHaveProperty('_isActive', false);
    });

    it('still has value', () => {
      expect(response.fields[fieldName]).toHaveProperty('value', value);
    });
  });

  describe('on destroy', () => {
    beforeEach(() => {
      act(() => {
        response.mountField({
          name: 'test',
          initialValue: 1234,
          initialValid: true
        });
        response.mountField({
          name: fieldName,
          initialValue: value,
          initialValid: false
        });
        response.unmountField({ name: fieldName, destroy: true });
      });
    });

    describe('field state', () => {
      it('matches snapshot', () => {
        expect(response.fields).toMatchSnapshot();
      });

      it('does not contain previous field', () => {
        expect(response.fields[fieldName]).toBe(undefined);
      });
    });
  });

  describe('on provided args', () => {
    beforeEach(() => {
      act(() =>
        response.mountField({
          name: 'test',
          initialError: 'some error',
          initialTouched: true,
          initialValue: 'hello',
          destroyOnUnmount: false,
          initialValid: true,
          validate: jest.fn(),
          validateOnBlur: false,
          validateOnChange: false,
          validateOnUpdate: true
        })
      );
    });

    describe('field state', () => {
      it('matches snapshot', () => {
        expect(response.fields).toMatchSnapshot();
      });
    });
  });
});

describe('on blur field', () => {
  const fieldName = 'TargetField';
  const validate = jest.fn();

  beforeEach(() => {
    mount(<Fixture />);
  });

  describe('field state', () => {
    beforeEach(() => {
      act(() => {
        response.mountField({ name: 'IgnoredField' });
        response.mountField({ name: fieldName, validate });
        response.blurField({ name: fieldName });
      });
    });

    it('matches snapshot', () => {
      expect(response.fields).toMatchSnapshot();
    });

    it('is touched', () => {
      expect(response.fields[fieldName]).toHaveProperty('touched', true);
    });
  });

  describe('with [validateOnBlur = true]', () => {
    beforeEach(() => {
      act(() => {
        response.mountField({ name: 'IgnoredField' });
        response.mountField({
          name: fieldName,
          validate,
          validateOnBlur: true,
          initialValid: true
        });
        response.blurField({ name: fieldName });
      });
    });
    describe('validate function', () => {
      it('is called', () => {
        expect(validate).toBeCalledTimes(1);
        expect(validate).toBeCalledWith(undefined, response.fields);
      });
    });
  });

  describe('with [validateOnBlur = false]', () => {
    describe('validate function', () => {
      beforeEach(() => {
        act(() => {
          response.mountField({ name: 'IgnoredField' });
          response.mountField({
            name: fieldName,
            validate,
            validateOnBlur: false
          });
          response.blurField({ name: fieldName });
        });
      });

      it('is not called', () => {
        expect(validate).toBeCalledTimes(0);
      });
    });
  });
});

describe('on change field', () => {
  const fieldName = 'TargetField';
  const value = 'Hello';
  const validate = jest.fn();

  beforeEach(() => {
    mount(<Fixture />);
  });

  describe('field value', () => {
    beforeEach(() => {
      act(() => {
        response.mountField({ name: 'IgnoredField' });
        response.mountField({ name: fieldName });
        response.setFieldValue({ name: fieldName, value });
      });
    });

    it('matches snapshot', () => {
      expect(response.fields).toMatchSnapshot();
    });

    it('has new value', () => {
      expect(response.fields[fieldName]).toHaveProperty('value', value);
    });
  });

  describe('with [validateOnChange = true]', () => {
    beforeEach(() => {
      act(() => {
        response.mountField({ name: 'IgnoredField' });
        response.mountField({
          name: fieldName,
          validate,
          validateOnChange: true,
          initialValid: true
        });
        response.setFieldValue({ name: fieldName, value });
      });
    });
    describe('validate function', () => {
      it('is called', () => {
        expect(validate).toBeCalledTimes(1);
        expect(validate).toBeCalledWith(value, response.fields);
      });
    });
  });

  describe('with [validateOnChange = false]', () => {
    describe('validate function', () => {
      beforeEach(() => {
        act(() => {
          response.mountField({ name: 'IgnoredField' });
          response.mountField({
            name: fieldName,
            validate,
            validateOnChange: false
          });
          response.blurField({ name: fieldName });
        });
      });

      it('is not called', () => {
        expect(validate).toBeCalledTimes(0);
      });
    });
  });
});

describe('on change field (callback)', () => {
  const fieldName = 'TargetField';
  const initialValue = 'old value';
  const value = 'Hello';
  const callback = jest.fn(() => value);

  beforeEach(() => {
    mount(<Fixture />);
  });

  describe('field value', () => {
    beforeEach(() => {
      act(() => {
        response.mountField({ name: 'IgnoredField' });
        response.mountField({ name: fieldName, initialValue });
        response.setFieldValue({ name: fieldName, value: callback });
      });
    });

    it('matches snapshot', () => {
      expect(response.fields).toMatchSnapshot();
    });

    it('has new value', () => {
      expect(response.fields[fieldName]).toHaveProperty('value', value);
    });

    it('calls callback with old value', () => {
      expect(callback).toBeCalledTimes(1);
      expect(callback).toHaveBeenCalledWith(initialValue);
    });
  });
});

describe('on validate field', () => {
  const fieldName = 'TargetField';
  const validate = jest.fn();
  const value = 'Field Value';

  beforeEach(() => {
    mount(<Fixture />);
  });

  describe('validation function', () => {
    beforeEach(() => {
      act(() => {
        response.mountField({ name: 'IgnoredField' });
        response.mountField({
          name: fieldName,
          validate,
          initialValid: true,
          initialValue: value
        });
        response.validateField({ name: fieldName });
      });
    });

    it('is called', () => {
      expect(validate).toBeCalledTimes(1);
      expect(validate).toBeCalledWith(value, {
        ...response.fields,
        [fieldName]: {
          ...response.fields[fieldName],
          isValid: true,
          error: undefined
        }
      });
    });
  });

  describe('on failed validation', () => {
    const error = 'Name must be longer';

    beforeEach(() => {
      validate.mockImplementation(() => {
        throw error;
      });
      act(() => {
        response.mountField({ name: 'IgnoredField' });
        response.mountField({
          name: fieldName,
          validate,
          initialValid: true,
          initialValue: value
        });
        response.validateField({ name: fieldName });
      });
    });

    describe('form state', () => {
      it('is not valid', () => {
        expect(response).toHaveProperty('isValid', false);
      });
    });

    describe('field state', () => {
      it('matches snapshot', () => {
        expect(
          Object.keys(response.fields).reduce(
            // Trim mock from snapshot
            (p, key) => ({
              ...p,
              [key]: { ...p[key], _validate: 'stub' as any }
            }),
            response.fields
          )
        ).toMatchInlineSnapshot(`
          Object {
            "IgnoredField": Object {
              "_isActive": true,
              "_validate": "stub",
              "_validateOnBlur": false,
              "_validateOnChange": true,
              "_validateOnUpdate": false,
              "error": undefined,
              "isValid": false,
              "isValidating": false,
              "name": "IgnoredField",
              "touched": false,
              "value": undefined,
            },
            "TargetField": Object {
              "_isActive": true,
              "_validate": "stub",
              "_validateOnBlur": false,
              "_validateOnChange": true,
              "_validateOnUpdate": false,
              "error": "Name must be longer",
              "isValid": false,
              "isValidating": false,
              "name": "TargetField",
              "touched": false,
              "value": "Field Value",
            },
          }
        `);
      });

      it('is not valid', () => {
        expect(response.fields[fieldName]).toHaveProperty('isValid', false);
      });

      it('has error value', () => {
        expect(response.fields[fieldName]).toHaveProperty('error', error);
      });
    });
  });

  describe('on successful validation', () => {
    beforeEach(() => {
      validate.mockReturnValue(undefined);
      act(() => {
        response.mountField({ name: 'IgnoredField', initialValid: true });
        response.mountField({
          name: fieldName,
          validate,
          initialValid: true,
          initialValue: value
        });
        response.validateField({ name: fieldName });
      });
    });

    describe('form state', () => {
      it('is valid', () => {
        expect(response).toHaveProperty('isValid', true);
      });
    });

    describe('field state', () => {
      it('matches snapshot', () => {
        expect(
          Object.keys(response.fields).reduce(
            // Trim mock from snapshot
            (p, key) => ({
              ...p,
              [key]: { ...p[key], _validate: 'stub' as any }
            }),
            response.fields
          )
        ).toMatchInlineSnapshot(`
          Object {
            "IgnoredField": Object {
              "_isActive": true,
              "_validate": "stub",
              "_validateOnBlur": false,
              "_validateOnChange": true,
              "_validateOnUpdate": false,
              "error": undefined,
              "isValid": true,
              "isValidating": false,
              "name": "IgnoredField",
              "touched": false,
              "value": undefined,
            },
            "TargetField": Object {
              "_isActive": true,
              "_validate": "stub",
              "_validateOnBlur": false,
              "_validateOnChange": true,
              "_validateOnUpdate": false,
              "error": undefined,
              "isValid": true,
              "isValidating": false,
              "name": "TargetField",
              "touched": false,
              "value": "Field Value",
            },
          }
        `);
      });

      it('is valid', () => {
        expect(response.fields[fieldName]).toHaveProperty('isValid', true);
      });

      it('has no error value', () => {
        expect(response.fields[fieldName]).toHaveProperty('error', undefined);
      });
    });
  });
});

describe('on validate field (async)', () => {
  const fieldName = 'TargetField';
  let resolve: (a?: any) => void;
  let reject: (a?: any) => void;

  beforeEach(() => {
    mount(<Fixture />);
  });

  beforeEach(() => {
    act(() => {
      response.mountField({ name: 'IgnoredField' });
      response.mountField({
        name: fieldName,
        validate: () =>
          new Promise((res, rej) => {
            resolve = res;
            reject = rej;
          })
      });
    });
  });

  describe('on call', () => {
    beforeEach(() => {
      act(() => {
        response.validateField({ name: fieldName });
      });
    });

    describe('field state', () => {
      it('is validating', () => {
        expect(response.fields[fieldName]).toHaveProperty('isValidating', true);
      });
    });
  });

  describe('on resolve', () => {
    beforeEach(async () => {
      act(() => {
        response.validateField({ name: fieldName });
      });
      await act(async () => {
        resolve();
      });
    });

    describe('field state', () => {
      it('matches snapshot', () => {
        expect(response.fields[fieldName]).toMatchSnapshot();
      });

      it('is not validating', () => {
        expect(response.fields[fieldName]).toHaveProperty(
          'isValidating',
          false
        );
      });

      it('has no error', () => {
        expect(response.fields[fieldName]).toHaveProperty('error', undefined);
      });

      it('is valid', () => {
        expect(response.fields[fieldName]).toHaveProperty('isValid', true);
      });
    });
  });

  describe('on resolve (with value)', () => {
    const val = 'Some error';

    beforeEach(async () => {
      act(() => {
        response.validateField({ name: fieldName });
      });
      await act(async () => {
        resolve(val);
      });
    });

    describe('field state', () => {
      it('matches snapshot', () => {
        expect(response.fields[fieldName]).toMatchSnapshot();
      });

      it('is not validating', () => {
        expect(response.fields[fieldName]).toHaveProperty(
          'isValidating',
          false
        );
      });

      it('has error', () => {
        expect(response.fields[fieldName]).toHaveProperty('error', val);
      });

      it('is not valid', () => {
        expect(response.fields[fieldName]).toHaveProperty('isValid', false);
      });
    });
  });

  describe('on reject (with value)', () => {
    const val = 'Some error';

    beforeEach(async () => {
      act(() => {
        response.validateField({ name: fieldName });
      });
      await act(async () => {
        reject(val);
      });
    });

    describe('field state', () => {
      it('matches snapshot', () => {
        expect(response.fields[fieldName]).toMatchSnapshot();
      });

      it('is not validating', () => {
        expect(response.fields[fieldName]).toHaveProperty(
          'isValidating',
          false
        );
      });

      it('has error', () => {
        expect(response.fields[fieldName]).toHaveProperty('error', val);
      });

      it('is not valid', () => {
        expect(response.fields[fieldName]).toHaveProperty('isValid', false);
      });
    });
  });

  describe('on multiple validation calls', () => {
    const val = 'Some error';

    beforeEach(() => {
      act(() => {
        response.validateField({ name: fieldName });
      });
    });

    describe('on old resolves -> new pending', () => {
      beforeEach(async () => {
        const initialResolve = resolve;
        act(() => {
          response.validateField({ name: fieldName });
        });

        await act(async () => {
          initialResolve(val);
        });
      });

      describe('field state', () => {
        it('is validating', () => {
          expect(response.fields[fieldName]).toHaveProperty(
            'isValidating',
            true
          );
        });
      });
    });

    describe('on new resolves -> old rejects', () => {
      beforeEach(async () => {
        const initialResolve = resolve;
        act(() => {
          response.validateField({ name: fieldName });
        });

        await act(async () => {
          resolve();
          initialResolve(val);
        });
      });

      describe('field state', () => {
        it('is not validating', () => {
          expect(response.fields[fieldName]).toHaveProperty(
            'isValidating',
            false
          );
        });

        it('has no error', () => {
          expect(response.fields[fieldName]).toHaveProperty('error', undefined);
        });

        it('is valid', () => {
          expect(response.fields[fieldName]).toHaveProperty('isValid', true);
        });
      });
    });
  });
});

describe('on validate fields', () => {
  const validate1 = jest.fn();
  const validate2 = jest.fn();

  beforeEach(() => {
    mount(<Fixture />);
    act(() => {
      response.mountField({ name: 'IgnoredField' });
      response.mountField({ name: 'field1', validate: validate1 });
      response.mountField({ name: 'field2', validate: validate2 });
      response.validateFields();
    });
  });

  describe('all validation functions', () => {
    it('are called', () => {
      expect(validate1).toBeCalledTimes(1);
      expect(validate2).toBeCalledTimes(1);
    });
  });
});
