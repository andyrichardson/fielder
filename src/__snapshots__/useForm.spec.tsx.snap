// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial call matches snapshot 1`] = `
Object {
  "blurField": [Function],
  "fields": Object {},
  "isValid": true,
  "isValidating": false,
  "mountField": [Function],
  "setFieldValue": [Function],
  "unmountField": [Function],
  "validateField": [Function],
  "validateFields": [Function],
}
`;

exports[`on blur field field state matches snapshot 1`] = `
Object {
  "IgnoredField": Object {
    "_isActive": true,
    "_validate": undefined,
    "_validateOnBlur": false,
    "_validateOnChange": true,
    "_validateOnUpdate": false,
    "error": undefined,
    "isValid": false,
    "isValidating": false,
    "name": "IgnoredField",
    "touched": false,
    "value": undefined,
  },
  "TargetField": Object {
    "_isActive": true,
    "_validate": [MockFunction],
    "_validateOnBlur": false,
    "_validateOnChange": true,
    "_validateOnUpdate": false,
    "error": undefined,
    "isValid": false,
    "isValidating": false,
    "name": "TargetField",
    "touched": true,
    "value": undefined,
  },
}
`;

exports[`on change field (callback) field value matches snapshot 1`] = `
Object {
  "IgnoredField": Object {
    "_isActive": true,
    "_validate": undefined,
    "_validateOnBlur": false,
    "_validateOnChange": true,
    "_validateOnUpdate": false,
    "error": undefined,
    "isValid": false,
    "isValidating": false,
    "name": "IgnoredField",
    "touched": false,
    "value": undefined,
  },
  "TargetField": Object {
    "_isActive": true,
    "_validate": undefined,
    "_validateOnBlur": false,
    "_validateOnChange": true,
    "_validateOnUpdate": false,
    "error": undefined,
    "isValid": false,
    "isValidating": false,
    "name": "TargetField",
    "touched": false,
    "value": "Hello",
  },
}
`;

exports[`on change field field value matches snapshot 1`] = `
Object {
  "IgnoredField": Object {
    "_isActive": true,
    "_validate": undefined,
    "_validateOnBlur": false,
    "_validateOnChange": true,
    "_validateOnUpdate": false,
    "error": undefined,
    "isValid": false,
    "isValidating": false,
    "name": "IgnoredField",
    "touched": false,
    "value": undefined,
  },
  "TargetField": Object {
    "_isActive": true,
    "_validate": undefined,
    "_validateOnBlur": false,
    "_validateOnChange": true,
    "_validateOnUpdate": false,
    "error": undefined,
    "isValid": false,
    "isValidating": false,
    "name": "TargetField",
    "touched": false,
    "value": "Hello",
  },
}
`;

exports[`on mount field on default args field state matches snapshot 1`] = `
Object {
  "test": Object {
    "_isActive": true,
    "_validate": undefined,
    "_validateOnBlur": false,
    "_validateOnChange": true,
    "_validateOnUpdate": false,
    "error": undefined,
    "isValid": false,
    "isValidating": false,
    "name": "test",
    "touched": false,
    "value": undefined,
  },
}
`;

exports[`on mount field on provided args field state matches snapshot 1`] = `
Object {
  "test": Object {
    "_isActive": true,
    "_validate": [MockFunction],
    "_validateOnBlur": false,
    "_validateOnChange": false,
    "_validateOnUpdate": true,
    "error": "some error",
    "isValid": true,
    "isValidating": false,
    "name": "test",
    "touched": true,
    "value": "hello",
  },
}
`;

exports[`on unmount field fields state matches snapshot 1`] = `
Object {
  "TargetField": Object {
    "_isActive": false,
    "_validate": undefined,
    "_validateOnBlur": false,
    "_validateOnChange": true,
    "_validateOnUpdate": false,
    "error": undefined,
    "isValid": false,
    "isValidating": false,
    "name": "TargetField",
    "touched": false,
    "value": 5678,
  },
  "test": Object {
    "_isActive": true,
    "_validate": undefined,
    "_validateOnBlur": false,
    "_validateOnChange": true,
    "_validateOnUpdate": false,
    "error": undefined,
    "isValid": true,
    "isValidating": false,
    "name": "test",
    "touched": false,
    "value": 1234,
  },
}
`;

exports[`on unmount field on destroy field state matches snapshot 1`] = `
Object {
  "test": Object {
    "_isActive": true,
    "_validate": undefined,
    "_validateOnBlur": false,
    "_validateOnChange": true,
    "_validateOnUpdate": false,
    "error": undefined,
    "isValid": true,
    "isValidating": false,
    "name": "test",
    "touched": false,
    "value": 1234,
  },
}
`;

exports[`on unmount field on provided args field state matches snapshot 1`] = `
Object {
  "test": Object {
    "_isActive": true,
    "_validate": [MockFunction],
    "_validateOnBlur": false,
    "_validateOnChange": false,
    "_validateOnUpdate": true,
    "error": "some error",
    "isValid": true,
    "isValidating": false,
    "name": "test",
    "touched": true,
    "value": "hello",
  },
}
`;

exports[`on validate field (async) on reject (with value) field state matches snapshot 1`] = `
Object {
  "_isActive": true,
  "_validate": [Function],
  "_validateOnBlur": false,
  "_validateOnChange": true,
  "_validateOnUpdate": false,
  "error": "Some error",
  "isValid": false,
  "isValidating": false,
  "name": "TargetField",
  "touched": false,
  "value": undefined,
}
`;

exports[`on validate field (async) on resolve (with value) field state matches snapshot 1`] = `
Object {
  "_isActive": true,
  "_validate": [Function],
  "_validateOnBlur": false,
  "_validateOnChange": true,
  "_validateOnUpdate": false,
  "error": "Some error",
  "isValid": false,
  "isValidating": false,
  "name": "TargetField",
  "touched": false,
  "value": undefined,
}
`;

exports[`on validate field (async) on resolve field state matches snapshot 1`] = `
Object {
  "_isActive": true,
  "_validate": [Function],
  "_validateOnBlur": false,
  "_validateOnChange": true,
  "_validateOnUpdate": false,
  "error": undefined,
  "isValid": true,
  "isValidating": false,
  "name": "TargetField",
  "touched": false,
  "value": undefined,
}
`;
